### PR TITLE
[SPARK-10121] [SQL] Thrift server always use the latest class loader provided by the conf of executionHive's state

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-10121

Looks like the problem is that if we add a jar through another thread, the thread handling the JDBC session will not get the latest classloader.

### CODE REVIEW

**Comment by yhuai** [comment]
@liancheng I manually tested it and this looks like a right fix. How to write a unit test? Basically, I need to start the thriftserver and add a jar from another thread (e.g. spark shell) and test it in a jdbc connection.

**Comment by liancheng** [comment]
This can be tested using `HiveThriftJdbcTest.withMultipleConnectionJdbcStatement()`. This method accepts zero or more `Statement => Unit` functions, and runs them within different JDBC sessions. Although these functions are executed sequentially, all the connections are established at the beginning, so different function are guaranteed to be executed in different thread.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 1]
```diff
@@ -377,6 +377,60 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
+          }
+          expectedResult.close()
+
+          assert(expectedResultBuffer === actualResultBuffer)
+
+          statement.executeQuery("DROP TABLE IF EXISTS addJar")
+          statement.executeQuery("DROP TABLE IF EXISTS smallKV")
+      }
+    )
+  }
```
Comment (yhuai) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: @liancheng is this test good?

Comment (liancheng) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Yeah, this LGTM.

**Comment by liancheng** [comment]
LGTM. Merging to master and branch-1.5.

**Pull request closed without merging**